File: app/src/main/java/com/abitcreative/popularmovies/async/NetworkAsync.java
package com.abitcreative.popularmovies.async;

import android.os.AsyncTask;
import android.util.Log;

import java.io.IOException;

import retrofit2.Call;

/**
 * Created by daber on 19/10/16.
 */

public class NetworkAsync<T> extends AsyncTask<Void, Void, T> {
    public interface OnNetworkResponseListener<T> {
        void onNetworkResponse(T response);
    }

    private final static String TAG = NetworkAsync.class.getSimpleName();

    private OnNetworkResponseListener<T> listener;
    private Call<T> call;

    public NetworkAsync(Call<T> call, OnNetworkResponseListener<T> listener) {
        this.listener = listener;
        this.call = call;
    }

    @Override
    protected T doInBackground(Void... params) {
        try {
            T t = call.execute().body();
            return t;
        } catch (IOException e) {
            Log.e(TAG, "Fetching data failed", e);
            return null;
        }
    }

    @Override
    protected void onPostExecute(T t) {
        listener.onNetworkResponse(t);
    }
}

File: app/src/main/java/com/abitcreative/popularmovies/async/NetworkAsync.kt
package com.abitcreative.popularmovies.async

import android.os.AsyncTask
import android.util.Log
import retrofit2.Call

/**
 * Created by daber on 15/10/16.
 */
open class NetworkAsync<T>(val call: Call<T>, val onResult: (T?) -> Unit) : AsyncTask<Void, Void, T?>() {
    val TAG = javaClass.simpleName
    override fun doInBackground(vararg params: Void?): T? {
        try {
            return call.execute().body()
        } catch (e: Exception) {
            Log.e(TAG, "Exception while making request", e)
            return null
        }
    }

    override fun onPostExecute(result: T?) {
        onResult(result)
    }

}